src/main/scala/linting/rule/RenameDesiredNames.scala

```diff
@@ -0,0 +1,197 @@
+    case other => other.mapStmt(onStmt(moduleNameMap))
+  }
+
+  /** Renames the modules in a circuit given a mapping of old names to new names
+    *
+    * @param nameMappings mapping of old to new names
+    * @param circuit the circuit to rename
+    */
+  def apply(nameMappings: Map[String, String], circuit: Circuit): Circuit = {
+    val modules = circuit.modules.map {
+      case mod: Module => mod.mapStmt(onStmt(nameMappings)).mapString(m => nameMappings.getOrElse(m, m))
+      case ext: ExtModule => ext
+    }
+    val main = nameMappings.getOrElse(circuit.main, circuit.main)
+    circuit.copy(main = main, modules = modules)
+  }
+}
+
+/** A strategy for generating a new names for modules
+  */
+sealed trait NamingStrategy {
+
+  /** Generates a new stable module name based on the desired name and the module IR node
+    *
+    * called by [[LintConflictingModuleNames]] to rename modules
+    *
+    * @param desiredName the requested name for the module, generated name should contain this name
+    * @param module the module targeted by the desiredName
+    */
+  def getName(desiredName: String)(module: Module): String
+}
+
+/** A naming strategy that always returns the desired name
+  */
+case object ExactNamingStrategy extends NamingStrategy {
+  def getName(desiredName: String)(module: Module): String = desiredName
+}
+
+/** Specifies a naming strategy to use for a module and the modules that it collides with
+  */
+case class NamingStrategyAnnotation(
+  strategy: NamingStrategy,
+  target: IsModule
+) extends SingleTargetAnnotation[IsModule] with HasSerializationHints {
+  def typeHints: Seq[Class[_]] = Seq(
+    ExactNamingStrategy.getClass
+  )
+
+  def duplicate(newTarget: IsModule): NamingStrategyAnnotation = {
+    this.copy(target = newTarget)
+  }
+}
+
+/** Specifies the desired name to rename the module to, overriding the Module.desiredName
+  */
+case class OverrideDesiredNameAnnotation(
+  desiredName: String,
+  target: IsModule
+) extends SingleTargetAnnotation[IsModule] {
+  def duplicate(newTarget: IsModule): OverrideDesiredNameAnnotation = {
+    this.copy(target = newTarget)
+  }
+}
+
+/** Renames modules based on their overridden desired names
+  *
+  * Desired module name overrides are specified by
+  * [[OverrideDesiredNameAnnotation]]. If one of the modules is targeted by a
+  * [[NamingStrategyAnnotation]] then the transform will attempt to rename
+  * modules according to that naming strategy. The default naming strategy is
+  * [[ExactNamingStrategy]] which will throw an exception if there are any
+  * module name conflicts.
+  */
+class RenameDesiredNames extends Transform with DependencyAPIMigration {
+
+  override def prerequisites = Seq(Dependency[DedupModules])
+  override def optionalPrerequisiteOf = Seq(Dependency[VerilogEmitter])
+
+  override def invalidates(transform: Transform) = false
+
+  private def checkStrategy(
```

Comment:
I think we should document that if the behavior here is to return `None` if the strategy fails to result in unique names for all the modules using this strategy, as that was confusing to me